Respect "validate" callbacks

afn opened the pull request:
Currently, `before_validation` callbacks are invoked, but `validate` ones are not. This fixes #409.

afn commented on lib/activerecord-import/import.rb:
```diff
@@ -37,18 +36,21 @@ def valid_model?(model)
-      else
-        model.send(:run_callbacks, :validation, &validation_proc)
+      model.run_callbacks(:validation) do
+        if defined?(ActiveSupport::Callbacks::Filters::Environment) # ActiveRecord >= 4.1
```
Is this safe? Or is there any possibility that `ActiveSupport::Callbacks::Filters::Environment` is not defined because it has not yet been auto-loaded?

jkowens replied:
I'm not entirely certain, but I think that would be safe to assume.

jkowens commented on lib/activerecord-import/import.rb:
```diff
@@ -37,18 +36,21 @@ def valid_model?(model)
-      validation_proc = lambda do
-        @validators.each do |v|
-          if validation_context == v.options.fetch(:on, validation_context)
-            v.validate(model) if validate?(v, model)
```
I think you can delete the `#validate?` method below now.

afn replied:
Done

jkowens commented:
@afn thanks for working on a fix for this! I'm impressed that you figured out some of the gory validation requirements for the different versions of AR. Could you take a look at the error for AR 5.1? I just committed a change so that it is no longer an allowed failure.

jkowens commented on lib/activerecord-import/import.rb:
```diff
@@ -24,48 +24,40 @@ def initialize(name, index)
+        model._validate_callbacks = validate_callbacks
+        model.class.__reset_runner(:validate)
+      elsif model.class.respond_to?(:_validate_callbacks=) # ActiveRecord < 5.1
+        model.class._validate_callbacks = validate_callbacks
```
I'm a little worried this might not be thread safe. Could that be an issue?

afn replied:
Yeah, you're right :frowning_face:

Let me rework this so it doesn't change anything at the class level. (Note that modifying the model *instance* isn't thread-safe either, but ActiveRecord is already thread-unsafe in this regard, so it's probably fine to assume that the same instance won't be simultaneously accessed from multiple threads.)

afn commented:
@jkowens I've reverted back to the earlier version, and added support for ActiveRecord 5.1. See comments in the code and let me know what you think. Thanks!

BTW, the build is failing in Travis for seemingly unrelated reasons.

jkowens commented:
@afn can you force this to run the build again? I had to update the Travis config due to some changes on their end.

jkowens commented:
Looks good, thanks again for digging into this issue. If you would squash all those commits now, I'll merge this into master.

afn commented:
Happy to squash if you prefer, but if you haven't tried it already, I'd recommend using GitHub's "squash and merge" feature. That way the PR will still show the original granular commits (useful if someone wants to drill down into the history of this change) but your master branch will only contain a single commit.

jkowens commented on lib/activerecord-import/import.rb:
```diff
@@ -24,48 +24,56 @@ def initialize(name, index)
+              # The alternative here would be to copy-and-paste the bulk of the
+              # ActiveSupport::Callbacks#run_callbacks method, which is undesirable if there's
+              # no real-world use case for it.
+              raise "The :validate callback chain contains an 'around' callback, which is unsupported" unless runner.final?
```
I like this 👍

jkowens commented:
That's awesome, I need to read up on the latest Github features!
